### DISCUSSION TITLE
[ISSUE] US ZIP CODES: RANGE ISSUES

### DISCUSSION THREAD

**Discussion by Anubiella** [saleor/saleor#10071] [Deployment]
Hi everyone,
always nice to be here!

We're facing an issue regarding Shipping Methods 
Postal Codes division, to be more precise:

we're working within US country, so within US States ZIP codes,
an image justo to get the idea.
Talking broadly we're talking about 5 digit long ZIP codes.

![United States ZIP Codes - All US Postal Codes by States ⋆ Naijaknowhow 28-6-2022 17-28-11](https://user-images.githubusercontent.com/31857092/176219206-7e593db2-17d6-4af9-aad0-d2191c618dc4.png)

What happen is that if you start to add as Added Postal Codes, ZIP ranges with codes like 06001 or
00501, Saleor store those values without the beginning zeros, as you can see.
 
![Shipping Methods - Ale Bike Wear 28-6-2022 17-32-19](https://user-images.githubusercontent.com/31857092/176220073-b5394b6a-d4f5-43f5-8833-b65d9a6b0e79.png)

Now it's not just a matter of aesthetics, the problem is that the calculation to show or not the shipping methods
in these situation is messed up.

So, for instance, the shipping methods with the previous added ZIP codes will be visible and usable also, for example, 
in the Hawaii (State which have a completely different ZIP range)

Have you ever run into this problem?

Thanks in advance, 
god bless

**Comment by timuric** [comment]
@patrys @maarcingebala any thoughts?

**Comment by korycins** [comment]
Hey @Anubiella,
Can you provide more details here? What version of Saleor do you use?
Does the second screenshot a list of added zip codes to a shipping method? I just tested it on main branch, and I am getting values with 0 at the beginning (for exclude and include).
![image](https://user-images.githubusercontent.com/23282005/176654521-27558638-9b07-4a7c-9b06-6c616e81acd5.png)

Do you use the dashboard to upload the code range manually? OR API + some scripts?

**Reply by Anubiella** [reply]
Hi, thanks for the reply.

So, basically versione status is
![Configuration - Ale Bike Wear 1-7-2022 08-51-18](https://user-images.githubusercontent.com/31857092/176840247-97886fe1-016c-4b39-8476-0993aed655ff.png)

Yes, we manually add and set Postal Codes into a shipping method.
I've applied the zip codes and re-written forcing zeros before the code reaching 5 digits and that's ok,
but I still got weird response matching with the actual shipping address zip code,
I've tried with an extreme hawaian address (zip code like 9xxxx):
![Alé Bike Wear 1-7-2022 08-54-44](https://user-images.githubusercontent.com/31857092/176840806-8e52f451-a1f3-4fbf-b135-a913482efbf1.png)
 
 But as an available shipping methods we still see Team Area1
![Alé Bike Wear 1-7-2022 08-55-48](https://user-images.githubusercontent.com/31857092/176840983-af5e94f4-ea41-458f-b97b-dbd050b079aa.png)
 
 that should be forbidden because the hawaian zip code is 9xxxx that's outside of the boundaries set into the Team Area1
 shipping method
 
![Shipping Methods - Ale Bike Wear 1-7-2022 08-57-41](https://user-images.githubusercontent.com/31857092/176841441-263a5970-f1bf-46d6-ac42-a9de34f45508.png)
![Shipping Methods - Ale Bike Wear 1-7-2022 08-57-56](https://user-images.githubusercontent.com/31857092/176841462-68b11597-bba5-4323-9475-ff07966c1ac3.png)
![Shipping Methods - Ale Bike Wear 1-7-2022 08-58-17](https://user-images.githubusercontent.com/31857092/176841477-2590d027-b840-432b-96a6-a6fa2373ebeb.png)
 

Is it an issue with the zip range applied or is it a bug/misunderstanding?

Thanks a lot in advance,
Lorenzo

**Comment by korycins** [comment]
Hey @Anubiella ,

Not sure what is wrong here. I just tried to reproduce it on the version that you provided. 
I've used the same address as you. I have one shipping method attached and added some postal rules.

The postal codes configuration:
![image](https://user-images.githubusercontent.com/23282005/176853452-454667d7-37e4-4c31-9c40-df98b24587e3.png)

The results that I am receiving after setting it up:
![image](https://user-images.githubusercontent.com/23282005/176853533-45db8935-d17d-480e-b114-32f046820b44.png)

Adding range 90000-99000 to include postal codes, gives me the result:
![image](https://user-images.githubusercontent.com/23282005/176853703-15ee01c3-5874-48bc-849a-d25d9ba715b5.png)

Removing all include postal codes rules, gives me the result:
![image](https://user-images.githubusercontent.com/23282005/176853873-43a83f52-b58e-46ec-a4f4-730430d1abca.png)

Adding 90000-99000 to exclude postal code gives me the result:
![image](https://user-images.githubusercontent.com/23282005/176854114-e69c95f0-b2f5-4731-8346-bdb9437f12fb.png)

Maybe you have multiple shipping zones where the same country is attached there?

**Reply by Anubiella** [reply]
Hi @korycins ,

thanks for the kind reply:

yeah, in the project there are several Shipping Zones that are based on the same Country (US),
by the way the Shipping Method I'm focusing on (TEAM AREA1) is merely on ONE Shipping Zone,
not on others, there are no mirroring issues I hope.

Maybe there are too many rules/codes?

And, above all, do you have by chance any advice to optimize the rules mass on a complex situation like the present one?

Thanks in advance

**Comment by timuric** [comment]
@Anubiella could you please share details about your use case?

**Reply by Anubiella** [reply]
Yes, or at least I'll try my best to explain the mess (my fellow colleague @DidoMarchet can underline or stress other points if
he finds something useful):
stakeholders needs to map different shipping zones in order to create macro-zones which are intended to handle
different users in the frontend (at least that was also our idea): 

b2b dealers, team customers in US, team customers in Canada, simple Canada customers, simple US customers and Puerto Rico.

So in our case there are active on US 3 shipping zones: dealers, simple customers and teams customers, each one
with its own shipping methods divided by price and zip codes.

Since is the main stakeholder that handle personally the zip codes (he knews which state/province zip codes are necessary
for the given method) I have not so many tools to analyze the situation.

What we've experimented is that, for example, for the Team Area 1, as soon as you start adding, and adding, and adding rules
over rules it starts to fail in filtering the method, but I cannot say no more unfortunately.

We'll continue to test those methods starting from the ground and adding rules with more attention,
trying to find the breaking point.

**Comment by timuric** [comment]
@Anubiella it sounds like you are better off adding your business logic to an excel file and using a webhook to programmatically to set shipping price and options.

https://docs.saleor.io/docs/3.x/developer/extending/apps/synchronous-webhooks#shipping-webhooks
